docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,505 @@
+    1. Sets the number of replicas on the control plane object from `spec.managed.controlPlane.replicas`.
+    1. Sets the k8s version on the control plane object from the `spec.managed.version`.
+    1. Add the following labels to the control plane object:
+       ```yaml
+        cluster.x-k8s.io/managed: <cluster-name>
+       ```
+    1. Creates the control plane object.
+    1. Sets the `spec.controlPlaneRef` field for the Cluster object.
+    1. Saves the cluster object in the API server.
+5. Machine deployment object creation
+    1. For each `spec.managed.worker.deployments` item in the list
+    1. Create a name `<cluster-name>-<worker-set-name>` (if too long, hash it)
+    1. Initializes a new MachineDeployment object.
+    1. Sets the `clusterName` field on the MD object
+    1. Sets the `replicas` field on the MD object using `replicas` field for the set of worker nodes.
+    1. Sets the `version` field on the MD object from the `spec.managed.version`.
+    1. Sets the `spec.template.spec.bootstrap` on the MD object from the `ClusterClass.spec.worker.deployments[i].template.bootstrap.ref` field.
+    1. Sets the `spec.template.spec.infrastructureRef` on the MD object from the `ClusterClass.spec.worker.deployments[i].template.infrastructure.ref` field.
+    1. Generates the set of labels to be set on the MD object. The labels are additive to the class' labels list, and the value in the `spec.managed.worker.deployments[i].labels` takes precedence over any set by the ClusterClass. Also, include the managed topology label stated below:
+       ```yaml
+        cluster.x-k8s.io/managed: <cluster-name>-<worker-set-name>
+       ```
+      Note: The managed topology label needs to be set on the individual Machine objects as well.
+    1. Creates the Machine Deployment object in the API server.
+
+![Creation of cluster with ClusterClass][./images/cluster-class/create.png]
+
+##### Update an existing Cluster using ClusterClass
+This section talks about updating a cluster which was created using a `ClusterClass` object.
+1. User updates the `cluster.spec.managed` field adhering to the update validation [criteria](#clusterclass-2).
+2. For the ControlPlane object in `spec.managed.controlPlane`, the cluster controller checks for the presence of the control plane object using the name `<cluster-name>`. If found,
+    1. Compares and updates the number of replicas, if necessary.
+    1. Compares and updates the k8s version, if necessary.
+    1. Updates the KCP object in the API server.
+3. The cluster controller reconciles the list of required machine deployments with the current list of managed machine deployments by:
+    1. Adding/Removing MachineDeployment if necessary.
+    1. Comparing and updating the number of replicas, if necessary.
+    1. Comparing and updating the k8s version for the MD, if necessary.
```

Comment:
This would result in all machineDeployments being upgraded at the same time. Are we considering any kind of policy/strategy e.g upgrading only one at a time and only move to next one when the previous one has completed, rollbacks... I think the upgrading workflow probably deserves its own section.

Reply:
I like the suggestion of having a control strategy around handling MD upgrades if a config (say k8s version) is changed. In the short term, we can use sensible defaults on the MachineDeploymentStrategy to ensure that not all the worker nodes get updated at the same. Currently, the default strategy type is `RollingUpdate` and `MaxSurge` is 1. This will ensure that even though all the MDs are getting updated at the same time, the overall availability of the worker nodes is not hampered.

Providing the observability metrics of the cluster status is outlined as a future goal and I suppose we can tackle the state of MD upgrades (yet to be started/in progress/completed) as a part of that enhancement.

Reply:
Another idea to control the update/rollback strategy could be to expose the `MachineDeploymentStrategy` field of the MD object under the `Managed` section of the Cluster object thereby not introducing a new level of complexity to these updates.

Reply:
We definitely need a rollout strategy for the class itself, although that can be added later

Reply:
Added a **Future goal** to implement an upgrade/rollback strategy.